xbmc/utils/FileUtils.cpp

```diff
@@ -162,3 +163,74 @@ bool CFileUtils::RemoteAccessAllowed(const std::string &strPath)
   }
   return false;
 }
+
+std::string CFileUtils::GetDateAdded(const std::string& strFileNameAndPath, const int& iAdvancedSettingsDateAdded)
+{
+  if (strFileNameAndPath.empty())
+  {
+    CLog::Log(LOGDEBUG, "%s empty strFileNameAndPath variable", __FUNCTION__);
+    return "";
+  }
+
+  try
+  {
+    std::string file = strFileNameAndPath;
+    if (URIUtils::IsStack(strFileNameAndPath))
+      file = CStackDirectory::GetFirstStackedFile(strFileNameAndPath);
+
+    if (URIUtils::IsInArchive(file))
+      file = CURL(file).GetHostName();
+
+    CDateTime dateAdded;
+    // Skip looking at the files ctime/mtime if defined by the user through as.xml
+    if (iAdvancedSettingsDateAdded > 0)
+    {
+      // Let's try to get the modification datetime
+      struct __stat64 buffer;
+      if (CFile::Stat(file, &buffer) == 0 && (buffer.st_mtime != 0 || buffer.st_ctime != 0))
+      {
+        time_t now = time(NULL);
+        time_t addedTime;
+        // Prefer the modification time if it's valid
+        if (iAdvancedSettingsDateAdded == 1)
+        {
+          if (buffer.st_mtime != 0 && (time_t)buffer.st_mtime <= now)
+            addedTime = (time_t)buffer.st_mtime;
+          else
+            addedTime = (time_t)buffer.st_ctime;
+        }
+        // Use the newer of the creation and modification time
+        else
+        {
+          addedTime = max((time_t)buffer.st_ctime, (time_t)buffer.st_mtime);
+          // if the newer of the two dates is in the future, we try it with the older one
+          if (addedTime > now)
+            addedTime = min((time_t)buffer.st_ctime, (time_t)buffer.st_mtime);
+        }
+
+        // make sure the datetime does is not in the future
+        if (addedTime <= now)
+        {
+          struct tm *time;
+#ifdef HAVE_LOCALTIME_R
+          struct tm result = {};
+          time = localtime_r(&addedTime, &result);
+#else
+          time = localtime(&addedTime);
+#endif
+          if (time)
+            dateAdded = *time;
+        }
+      }
+    }
+
+    if (!dateAdded.IsValid())
+      dateAdded = CDateTime::GetCurrentDateTime();
```

Comment:
as we have two different AS settings for video and music it's basically possible that both get out of sync because of their available values ;) .. IMO we can life with that.

I would do the following. Rename the method signature to `CDateTime CFileUtils::GetModificationDate(const std::string& strFileNameAndPath, bool& bUseLatestDate /* = true */)`. The parameter `bUseLatestDate` indicates to use the latest date out of the creation and modification date (same as AS value is larger 1).

In the database classes you add the AS magic. check which value is set and do the following.
* if value is 1, call to `CFileUtils::GetModificationDate()`
* if value is larger 1, call to `CFileUtils::GetModificationDate()` with `bUseLatestDate = false`
* if returned date is not valid use current date